### PR TITLE
Added test coverage for Yaku Functions and Hand Scoring Integration Testing

### PR DESCRIPTION
**Pull request by ChristopherJMiller**
**Changes**
- Added tests for all yaku and yakuman, under the Yaku test executable
- Added integration testing for hand score, under the Score_Hand executable (which should be expected in the future as more interesting cases are developed)
- Added `handformer.cpp`, which contains the function `Mahjong::HandFromNotation`. This can produce a list of pieces using [a common community notation](https://www.reddit.com/r/Mahjong/comments/dgth5z/is_there_a_standard_notation_for_tiles/) and is used heavily in testing for compactness

**Pull Request Goal**
Merge existing tests into mainline to allow bug fixing branches to branch from mainline instead of this feature branch.

**Note**
- This PR contains little fixing of what test failures were found, *these should be accomplished on specific patch PRs*.

### CODE REVIEW

File: include/handformer.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "piecetype.h"
+
+namespace Mahjong {
+auto HandFromNotation(std::string notation) -> std::vector<Piece>;
+}
```
Comment (aliceagain) [inline include/handformer.h:10]: nit: missing end of file new line

File: src/handformer.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          if (isHonor) {
+            return Piece(NOTATION_TO_HONOR[i]);
+          }
+          return Piece::formPiece(setSuit, i);
+        });
+      currentTiles.clear();
+    }
+  }
+  return result;
+}
```
Comment (aliceagain) [inline src/handformer.cpp:64]: nit: no new line

File: include/handformer.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#pragma once
```
Comment (aliceagain) [inline include/handformer.h:1]:
```
Warning: include-what-you-use reported diagnostics:

../../include/handformer.h should add these lines:
namespace Mahjong { class Piece; }

../../include/handformer.h should remove these lines:
- #include "piecetype.h"  // lines 6-6
```

File: src/handformer.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
+#include "handformer.h"
```
Comment (aliceagain) [inline src/handformer.cpp:1]:
```
../../src/handformer.cpp should add these lines:
#include <ctype.h>      // for isdigit, isalpha
#include <cstdint>      // for int8_t, uint8_t
#include <iterator>     // for back_insert_iterator, back_inserter

../../src/handformer.cpp should remove these lines:
- #include <iostream>  // lines 4-4
```

File: src/handformer.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  {6, Piece::GREEN_DRAGON},
+  {7, Piece::RED_DRAGON},
+};
+
+auto Mahjong::HandFromNotation(std::string notation) -> std::vector<Piece> {
+  std::vector<Piece> result;
+  std::vector<int8_t> currentTiles;
+  for (const auto& c : notation) {
+    if (std::isdigit(c)) {
+      auto value = c - '0';
```
Comment (aliceagain) [inline src/handformer.cpp:35]: This probably fine this stackoverflow answer points to this being safe with UTF-7 and UTF-8 and I don't think we need to support any additional encodings here. Consider a quick comment tho.

File: src/handformer.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      currentTiles.push_back(value);
+    } else if (!std::isdigit(c) && !std::isalpha(c)) {
+      currentTiles.push_back(-1);
```
Comment (aliceagain) [inline src/handformer.cpp:38]: Why do we push back a -1 here.

File: src/handformer.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    } else {
+      auto setSuit = Piece::ERROR;
+      auto isError = true;
```
Comment (aliceagain) [inline src/handformer.cpp:41]: There's a lot of nested if statements here and it's a bit confusing to read. Would you be able to un-nest some of the if statements? Maybe by using continue / return instead.

File: src/hands.cpp [hunk 8]
```diff
@@ -437,6 +437,7 @@ auto isRiichi(const GameState& state, int player, const std::vector<const Mahjon
     if (state.turnNum - state.hands.at(player).riichiRound <= ONEROUNDOFTURNS && state.lastCall < state.hands.at(player).riichiRound) {
```
Comment (aliceagain) [inline src/hands.cpp:437]: I would make a note of what needs to be changed here on the trello, but I think you shouldn't make these changes in this pr. Or you can just split this off to a seperate smaller cr.

File: src/piecetype.cpp [hunk 9]
```diff
@@ -65,6 +65,21 @@ auto Piece::fromWind(Wind w) -> Piece {
   }
 }
 
+auto Piece::formPiece(Piece::Type suit, uint8_t number) -> Piece {
+  auto valueAppend = number;
+  if (number == 1 || number == 9) {
```
Comment (aliceagain) [inline src/piecetype.cpp:70]: You can use isTerminal here

File: src/piecetype.cpp [hunk 10]
```diff
@@ -65,6 +65,21 @@ auto Piece::fromWind(Wind w) -> Piece {
...
+    valueAppend |= TERMINAL_BIT;
+  } else if (number == 0) {
```
Comment (aliceagain) [inline src/piecetype.cpp:72]: I really don't like this. Can we have an additional red five parameter?

File: src/piecetype.cpp [hunk 11]
```diff
@@ -65,6 +65,21 @@ auto Piece::fromWind(Wind w) -> Piece {
...
+    valueAppend = 5 | RED_FIVE;
+  }
+
+  if (suit == CHARACTER_SUIT || suit == PIN_SUIT || suit == BAMBOO_SUIT || suit == HONOR_SUIT) {
+    return Piece(suit | valueAppend);
+  } else {
```
Comment (aliceagain) [inline src/piecetype.cpp:78]: Don't need the else here.

File: tests/handformer.test.cpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  EXPECT_EQ(result, hand);
+}
+
+TEST(HandFormer, HandFromNotationInvalidPieceSet) {
+  std::vector<Piece> hand = {
+    Piece(Piece::ERROR),
+    Piece(Piece::ERROR),
+    Piece(Piece::ERROR)};
+  auto result = HandFromNotation("799g");
+  std::cout << result[0].toStr() << std::endl;
```
Comment (aliceagain) [inline tests/handformer.test.cpp:75]: Please remove both of the couts.

File: tests/handformer.test.cpp [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+  EXPECT_EQ(result, hand);
+}
```
Comment (aliceagain) [inline tests/handformer.test.cpp:77]: nit: /n

**Review by aliceagain — CHANGES_REQUESTED** [review]
I would prefer if we can split this into 2-4 PRs since it is so large. I know there's not a great way to break it up more, but in future we should aim for smaller more frequent PRs. I'd rather two 400 line PRs than one 800 line one. I think you could at least split this one into handformer, yakus, yakuman, scoring testing.

I when ahead and reviewed the handformer changes so you have some feedback and requested changes there.

Also as a general note we should post the output of the tests in our PRs. 

The Handformer module looks nice overall and will be helpful in future debugging as well. No longer have to type all the pieces out by hand lmao.

File: src/piecetype.cpp (hunk 9 above)
Comment (ChristopherJMiller) [reply src/piecetype.cpp:70]: is terminal only checks for whether the terminal bit is setup. As this 1 or 9 check is for setting the terminal bit itself, this means isTerminal cannot be used.

File: src/piecetype.cpp (hunk 10 above)
Comment (ChristopherJMiller) [reply src/piecetype.cpp:72]: It is a part of the community standard for 0 to represent red five, this should be discussed further

**Pull request closed without merging**